runtime/doc/dev_style.txt

```diff
@@ -41,6 +41,34 @@ All header files should start with `#pragma once` to prevent multiple inclusion.
         #pragma once
 <
 
+Headers system ~
+
+Neovim has uses two types of headers. There are "normal" headers and "defs"
+headers. Typically, each normal header will have a corresponding defs header,
+e.g. fileio.h and fileio_defs.h. This distinction is done to minimize
+recompilation of header files on change. Use the following rules to decide
+where to put what:
+
+- All non-symbols shall be moved to _defs.h:
+  - macros, i.e. `#define`.
+    - EXCEPTION: if the macro calls a function from a non-system header, then
+      it needs to be moved to a normal header.
+  - static inline functions, but only if its function declaration has a
+    REAL_FATTR_ALWAYS_INLINE attribute.
+  - typedefs
+  - structs
+  - enum constants
```

Comment:
Note: moving an enum constant (or `#define`d constant) to a _defs cannot lead to less recompilation on change if it can only be used by the callers of a specific function (ref #26622), as users of the enum constant must have already included the header for the function, and changes to the enum constant must be associated with changes to the function.